### ISSUE TITLE
Covox and Disney Sound Source

### ISSUE THREAD

**Issue by GranMinigun** [dosbox-staging/dosbox-staging#1120]
According to [this comment](https://github.com/dosbox-staging/dosbox-staging/issues/278#issuecomment-612655246), DOSBox actually emulates Covox Speech Thing and mistakenly names it as Disney Sound Source. It was suggested to fix documentation and configuration accordingly, and possibly incorporate emulation of DSS (PCem has it).

**Comment by dreamer** [comment]
Not sure if it's a "mistake". Covox ST and DSS are almost the same device - they had tiny implementation details and DSS came with a bad speaker, but otherwise perhaps there's no difference when it comes to emulation of both devices.

Video about both devices: https://www.youtube.com/watch?v=Wt4VcLF8Wug

Personally, I do **not** care about these so much, that I probably would welcome complete removal of the feature. There seems to be no benefit to using it when compared to SB. It's only a curiosity.

**Comment by volkertb** [comment]
I wouldn't call the differences between these two devices "tiny implementation details". The one was a dumb DAC, or even a simple resistor ladder, whereas the other had a 16-byte FIFO buffer that would stream sound to the DAC at a fixed rate of 7KHz. Also, the DSS could be auto-detected, whereas a Covox Speech Thing (and clones) could not. All of this made the two devices mutually incompatible, although a very few games employed a unified (suboptimal) way to drive both device types. A notable example of that is Another World (a.k.a. Out of This World).

That being said, it is indeed true that pretty much every game and application that supports either of these devices also supports Sound Blaster cards, and with sound quality that is at least as good, and often better. So I agree that there isn't any practical value in accurately emulating these parallel port audio devices. Unlike PCem (with respect to that impressive project), hardware emulation accuracy isn't relevant to DOSBox (Staging). Instead, the objective is an optimal DOS gaming experience on modern hardware.

**Comment by Wengier** [comment]
@volkertb You are right that hardware emulation accuracy isn't relevant to most DOSBox builds. However, it is very relevant to DOSBox-X, just like PCem. Both of them are trying to implement accurate emulation.

**Comment by kcgen** [comment]
@volkertb , 

Yes, I agree with you and @dreamer on the fidelity front: if the same DAC data is available at higher sampling rates or greater bit-depths, then that's always preferred (just as if an identical movie   recorded to VHS vs DVD vs BluRay vs 8K HEVC, will always look and sound best on the latter).

> , it is indeed true that pretty much every game and application that supports either of these devices also supports Sound Blaster (or better) cards

Are you aware of any games that support the CMS or DSS at a maximum? And therefore, emulating these property would still provide a better experience versus say, the PC speaker? 

> Staging's objective is an optimal DOS gaming experience on modern hardware.

Yes.. if DOSBox SVN's (and by extension our) implementation is currently broken, I'd rather we 1) fix them if we can import improvements from X or PCem (with all credit to those authors) or 2) if we cant fix them, then it would be better to temporarily disable them and hide their conf option(s) until we can properly support them.

**Comment by volkertb** [comment]
> Yes, I agree with you and @dreamer on the fidelity front: if the same DAC data is available at higher sampling rates or greater bit-depths, then that's always preferred (just as if an identical movie recorded to VHS vs DVD vs BluRay vs 8K HEVC, will always look and sound best on the latter).
> 

For this reason, there are a few sound cards other than the Sound Blaster 16 that *are* worth emulating in DOSBox (Staging), because they produce better quality sound in at least a few games:

- Gravis Ultrasound (obviously, due to stuff like hardware mixing, and DOSBox Staging already emulates this well)
- Sound Blaster Pro (because some older games supported stereo sound only on the SB Pro, and not on the SB16)
- Windows Sound System a.k.a. WSS (some games actually sounded the best with this option, because they supported OPL3 instead of OPL2 only with this card, or supported 16-bit audio only on this card)

You could even make the case to implement emulation for both the earlier revision of the Sound Blaster Pro (with dual OPL2 FM chips) and the later revision (with a single OPL3 FM chip), since there are supposedly some games that sound better on the one variant, and some that sound better on the other.

> Are you aware of any games that support the CMS or DSS at a maximum? And therefore, emulating these property would still provide a better experience versus say, the PC speaker?

In the case of the Covox Speech Thing (as well as an audio option labeled "MDO Intersound" in a few games, which apparently was compatible with it), there is an interesting thread here: https://www.vogons.org/viewtopic.php?f=62&t=57885&p=663383

TL;DR: there is a very small list of games that supported these dumb parallel port DACs, without supporting Sound Blaster. Notably a number of older games from Coktel Vision.

As for the Disney Sond Source, I don't know of any games that supported the DSS without also supporting Sound Blaster. As far as I know, even Disney, who distributed the Sound Source with a number of their games back in the day, still also supported Sound Blaster in those same games. (I'm not a 100% sure about that though. Perhaps someone else can chime in here.)

In the case of CMS, I believe there's a Vogons thread in which people are debating about a few games that supposedly sounded better (at least subjectively to some people) with CMS/Gameblaster music than with OPL2 FM music. One advantage of CMS/Gameblaster is that it supported stereo music from the start. Adlib and compatibles (with a single OPL2 chip, before the Sound Blaster Pro) did not support stereo music. There might even be a game or two that supported only CMS music and not Adlib/OPL2. Not sure, though. Also, CMS has more voices, if I'm not mistaken, even if the possible wave forms are much more limited compared to FM synthesis.

To be fair, the distinct CMS sound apparently has its fans, enough that people developed patches and drivers to make it work as a General MIDI device in games that otherwise don't support CMS. Also, Serdashop sells [an LPT device with CMS chips on it](https://www.serdashop.com/CMSLPT), to allow people to experience the unique CMS sound on computers lacking an ISA slot. :smile:

So basically, the CMS sounds very distinct and unique compared to FM synthesis, and would justify support for it for that reason.

in contrast, digital audio on a Covox Speech Thing or a Disney Sound Source instead of a Sound Blaster DAC wouldn't sound very distinct. At best it would sound similar, at worst it would just sound the same, but with for a lower playback quality. So not worth implementing.

> Yes.. if DOSBox SVN's (and by extension our) implementation is currently broken, I'd rather we 1) fix them if we can import improvements from X or PCem (with all credit to those authors) or 2) if we cant fix them, then it would be better to temporarily disable them and hide their conf option(s) until we can properly support them.

Hey, the licenses between all these projects are compatible (all GPL2, right?), so I'm all for copying the best existing solution (at least as a starting point) instead of reinventing the wheel. :slightly_smiling_face:

**Comment by kcgen** [comment]
> You could even make the case to implement emulation for both the earlier revision of the Sound Blaster Pro (with dual OPL2 FM chips) and the later revision (with a single OPL3 FM chip), since there are supposedly some games that sound better on the one variant, and some that sound better on the other.

Yes, DOSBox 0.74.x and SVN support these with `sbtype =  sb1, sb2, sbpro1, sbpro2, ...`

As for the CMS/GameBlaster, that's also supported: `sbtype = gb`

> Covox Speech Thing (as well as an audio option labeled "MDO Intersound"

Is there a best-of-the-best implementation out there? 
We certainly want to use that and give credit to those author(s)!

**Comment by Jkapp76** [comment]
I'm trying to play (last half of darkness) on dosbox-staging. It seems to use the Covox sound. The game exits with an error saying no coxox software loaded.

I'm new to the dss and Covox. Anyone have luck with this game.

**Comment by kcgen** [comment]
From @eulisker (thanks!)

> I have the game Alcatraz. On the box and in the PC Loading instructions (or whatever it should be called) it says it supports sound source

![IMG_1112](https://user-images.githubusercontent.com/1557255/162448716-77a553ec-d0d9-44cc-a42f-4f0bcbaa2c59.jpg)
![IMG_1111](https://user-images.githubusercontent.com/1557255/162448722-b79af548-c612-46f3-95d7-3bdc7013d32f.jpg)
[Alcatraz.zip](https://github.com/dosbox-staging/dosbox-staging/files/8452406/Alcatraz.zip)

(side note, MobyGames doesn't list Alcatraz in their DOS games that support the DSS: https://www.mobygames.com/attribute/sheet/attributeId,44/offset,0/p,2/so,0a/)

**Comment by kcgen** [comment]
Completed.

The `disney = true/false` conf setting is now:

```ini
[speaker]
lpt_dac = disney, covox, ston1, (or none/off)
```

This lets the use specify what type of parallel port DAC they'd like to use.
